lib/observability/tracing/client.go

```diff
@@ -226,3 +232,159 @@ func (f *RotatingFileClient) UploadTraces(ctx context.Context, protoSpans []*otl
 
 	return nil
 }
+
+// bufferedClient is an [otlptrace.Client] implementation that buffers uploaded
+// spans in memory until a real client and all resource attributes are available.
+type bufferedClient struct {
+	logger *slog.Logger
+
+	// mu synchronizes access to all below fields.
+	mu sync.Mutex
+
+	bufferedSpans [][]*otlp.ResourceSpans
+
+	// client and closer will not be modified and are safe to reference without
+	// locking after waitingForClient is false.
+	waitingForClient bool
+	client           otlptrace.Client
+	closer           io.Closer
+	started          bool
+
+	// delayedResourceAttrs will not be modified and is safe to reference
+	// without locking after waitingForDelayedResourceAttrs is false.
+	waitingForDelayedResourceAttrs bool
+	delayedResourceAttrs           []*otlpcommon.KeyValue
+}
+
+func newBufferedClient(cfg Config) (*bufferedClient, error) {
+	bufClient := &bufferedClient{
+		logger:                         cfg.Logger,
+		waitingForClient:               cfg.WaitForDelayedClient,
+		waitingForDelayedResourceAttrs: cfg.WaitForDelayedResourceAttrs,
+	}
+	if !cfg.WaitForDelayedClient {
+		var err error
+		bufClient.client, err = NewClient(cfg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return bufClient, nil
+}
+
+// Start implements [otlptrace.Client.Start].
+func (b *bufferedClient) Start(ctx context.Context) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.started = true
+	if b.client != nil {
+		return trace.Wrap(b.client.Start(ctx))
+	}
+	return nil
+}
+
+// Stop implements [otlptrace.Client.Stop].
+func (b *bufferedClient) Stop(ctx context.Context) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	var (
+		stopErr  error
+		closeErr error
+	)
+	if b.client != nil {
+		stopErr = trace.Wrap(b.client.Stop(ctx))
+	}
+	if b.closer != nil {
+		closeErr = trace.Wrap(b.closer.Close())
+	}
+	b.bufferedSpans = nil
+	return trace.NewAggregate(stopErr, closeErr)
+}
+
+// UploadTraces implements [otlptrace.Client.UploadTraces]. It may be called
+// concurrently. If the client is ready, the traces are immediately uploaded.
+// Otherwise, the traces are buffered in memory until the delayed client or
+// resource attributes are provided.
+func (b *bufferedClient) UploadTraces(ctx context.Context, protoSpans []*otlp.ResourceSpans) error {
+	b.mu.Lock()
+	if b.ready() {
+		b.mu.Unlock()
+		return b.uploadTracesToClient(ctx, protoSpans, b.delayedResourceAttrs)
+	}
+	b.bufferedSpans = append(b.bufferedSpans, protoSpans)
```

Comment:
Should we consider enforcing some kind of limit on the length of bufferSpans? I'm a little split on it - I think I particularly worry about some agent/bot getting stuck in the initial connection loop and this effectively leaking memory whilst stuck in that loop.

Reply:
yeah that's a valid concern, added a limit of 32 batches of spans (they get batched by `sdktrace.NewBatchSpanProcessor`)